res/controllers/midi-components-0.1.js

```diff
@@ -0,0 +1,1135 @@
+            return 1;
+        }
+    };
+
+    const HotcueButton = class extends Button {
+        constructor(options) {
+            if (options.number === undefined) {
+                throw "No hotcue number specified for HotcueButton constructor";
+            }
+            if (options.colorMapper !== undefined || options.sendRGB !== undefined) {
+                options.colorKey = "hotcue_" + options.number + "_color";
+            }
+            options.outKey = "hotcue_" + options.number + "_enabled";
+            super(options);
+        }
+        unshift() {
+            this.inKey = "hotcue_" + this.number + "_activate";
+        }
+        shift() {
+            this.inKey = "hotcue_" + this.number + "_clear";
+        }
+        output(value) {
+            const outval = this.outValueScale(value);
+            // NOTE: outputColor only handles hotcueColors
+            // and there is no hotcueColor for turning the LED
+            // off. So the `send()` function is responsible for turning the
+            // actual LED off.
+            if (this.colorKey !== undefined && outval !== this.off) {
+                this.outputColor(engine.getValue(this.group, this.colorKey));
+            } else {
+                this.send(outval);
+            }
+        }
+        outputColor(colorCode) {
+            // Sends the color from the colorCode to the controller. This
+            // method will not be called if no colorKey has been specified.
+            if (colorCode === undefined || colorCode < 0 || colorCode > 0xFFFFFF) {
+                console.warn("Ignoring invalid color code '" + colorCode + "' in outputColor()");
+                return;
+            }
+
+            if (this.colorMapper !== undefined) {
+                // This HotcueButton holds a reference to a ColorMapper. This means
+                // that the controller only supports a fixed set of colors, so we
+                // get the MIDI value for the nearest supported color and send it.
+                const nearestColorValue = this.colorMapper.getValueForNearestColor(colorCode);
+                this.send(nearestColorValue);
+            } else {
+                // Since outputColor has been called but no ColorMapper is
+                // available, we can assume that controller supports arbitrary
+                // RGB color output.
+                this.sendRGB(colorCodeToObject(colorCode));
+            }
+        }
+        sendRGB(_colorObject) {
+            // This method needs to be overridden in controller mappings,
+            // because the procedure is controller-dependent.
+            throw "sendRGB(colorObject) not implemented -- unable to send RGB colors";
+        }
+        connect() {
+            super.connect();
+            if (this.group !== undefined && this.colorKey !== undefined) {
+                // TODO (Swiftb0y): replace with arrow function once https://bugreports.qt.io/browse/QTBUG-95677 got fixed
```

Comment:
It's better not to assign ownership of TODOs so anyone can feel welcome to address them

```suggestion
                // TODO (https://bugreports.qt.io/browse/QTBUG-95677): replace with arrow function once this QT bug is fixed.
```

Reply:
I was not aware that this syntax was addressing anyone responsible for fixing things. I thought it was rather a signature so if someone wants to address the issue, they know who to talk to. Though to be fair, `git blame` would work just as well for that.

Reply:
Oh interesting!  This is a minor point so no big whoop.  (I've also found attaching specific bugs to TODOs makes them more likely to get fixed)